src/merkle/store/mod.rs

```diff
@@ -244,6 +244,75 @@ impl MerkleStore {
         })
     }
 
+    /// Reconstructs a path from the root until a leaf or empty node and returns its depth.
+    ///
+    /// The `tree_depth` argument will define up to what value the traverse will be performed. The
+    /// maximum value is [u64::BITS].
+    ///
+    /// The traversed path from leaf to root will start at the least significant bit of `index`,
+    /// and will be executed for `tree_depth` bits.
+    ///
+    /// # Errors
+    /// Returns an error if the provided root is not found, or if the path opens to a tree with
+    /// depth greater than `tree_depth`.
+    pub fn get_leaf_depth(
+        &self,
+        root: Word,
+        tree_depth: u8,
+        index: u64,
+    ) -> Result<u8, MerkleError> {
+        if tree_depth > 64 {
+            return Err(MerkleError::DepthTooBig(tree_depth as u64));
+        }
+
+        // it's not illegal to have a maximum depth of `0`; we should just return the root in that
+        // case. this check will simplify the implementation as we could overflow bits for depth
+        // `0`.
+        if tree_depth == 0 {
+            return Ok(0);
+        }
+
+        // check if the root exists, providing the proper error report if it doesn't
+        let empty = EmptySubtreeRoots::empty_hashes(tree_depth);
+        let mut hash: RpoDigest = root.into();
+        self.nodes
+            .get(&hash)
+            .ok_or(MerkleError::RootNotInStore(hash.into()))?;
+
+        // we traverse from root to leaf, so the path is reversed
+        let mut path = (index << (64 - tree_depth)).reverse_bits();
```

Comment:
I wonder if we should put a check somewhere above to make sure that `index` is not too big for the specified `tree_depth`.

Reply:
A couple of things to consider there

If we don't check this, we could use a normal, full `64` bits path, and just pass a tree depth of `8`. This will truncate the path to `8` and check from the leaf these 8 bits. However, this would halt at the check of "maximum depth" as this particular branch would continue up to the depth of `64` with the `DepthTooBig` error.

Having that said, if we check that maximum depth is really a leaf (i.e. no children), then it makes sense to validate the index. I'm adding the check.